docs/source/more_types.rst

```diff
@@ -0,0 +1,553 @@
+More types
+==========
+
+This section introduces a few additional kinds of types, including ``NoReturn``,
+``NewType``, ``TypedDict``, and types for async code. All of these are only
+situationally useful, so feel free to skip this section and come back when you
+have a need for some of them.
+
+Here's a quick summary of what's covered here:
+
+* ``NoReturn`` lets you tell mypy that a function never returns normally.
+
+* ``NewType`` lets you define a variant of a type that is treated as a
+  separate type by mypy but is identical to the original type at runtime.
+  For example, you can have ``UserId`` as a variant of ``int`` that is
+  just an ``int`` at runtime.
+
+* ``TypedDict`` lets you give precise types for dictionaries that represent
+  objects with a fixed schema, such as ``{'id': 1, 'items': ['x']}``.
+
+* Async types let you type check programs using async and await.
+
+.. _noreturn:
+
+The NoReturn type
+*****************
+
+Mypy provides support for functions that never return. For
+example, a function that unconditionally raises an exception:
+
+.. code-block:: python
+
+   from mypy_extensions import NoReturn
+
+   def stop() -> NoReturn:
+       raise Exception('no way')
+
+Mypy will ensure that functions annotated as returning ``NoReturn``
+truly never return, either implicitly or explicitly. Mypy will also
+recognize that the code after calls to such functions is unreachable
+and will behave accordingly:
+
+.. code-block:: python
+
+   def f(x: int) -> int:
+       if x == 0:
+           return x
+       stop()
+       return 'whatever works'  # No error in an unreachable block
+
+Install ``mypy_extensions`` using pip to use ``NoReturn`` in your code.
```

Comment:
NoReturn is actually in `typing_extensions`, so I don't think we should recommend using `mypy_extensions`. In recent versions, it's also in `typing`.